File: packages/browser-session-mcp/src/logs.ts
import { promises as fs, createReadStream } from "node:fs";
import path from "node:path";
import readline from "node:readline";

export const DEFAULT_LOGS_DIR = "/var/lib/browser-session-mcp/logs";

const SEQ_PAD = 5; // 99999 visits per session — plenty.

export type ConsoleEntry = {
  readonly t: string;
  readonly kind: "console";
  readonly type: string;
  readonly text: string;
};

export type NetworkEntry = {
  readonly t: string;
  readonly kind: "network";
  readonly method: string;
  readonly url: string;
  readonly status: number | null;
  readonly failure?: string;
};

export type VisitHeader = {
  readonly kind: "visit";
  readonly seq: number;
  readonly targetId: string;
  readonly url: string;
  readonly openedAt: string;
};

export type LogEntry = ConsoleEntry | NetworkEntry;

const sanitize = (s: string): string => s.replace(/[^A-Za-z0-9_-]/g, "_");
const pad = (n: number): string => String(n).padStart(SEQ_PAD, "0");

const sessionDir = (logsDir: string, sessionId: string): string =>
  path.join(logsDir, sanitize(sessionId));

const visitFile = (
  logsDir: string,
  sessionId: string,
  seq: number,
  targetId: string,
): string =>
  path.join(
    sessionDir(logsDir, sessionId),
    `${pad(seq)}-${sanitize(targetId)}.ndjson`,
  );

const seqFromFilename = (name: string): number => {
  const head = name.split("-", 1)[0];
  return Number.parseInt(head ?? "", 10);
};

/**
 * Per-visit NDJSON writer.
 *
 * Layout:
 *   <logsDir>/<sessionId>/<seq>-<targetId>.ndjson
 *
 * Each file starts with a {"kind":"visit",...} header. Subsequent lines are
 * console + network events. A new file is opened on every top-level
 * navigation (driven by listener.ts). The session folder is rm'd on close.
 */
export class LogWriter {
  private readonly logsDir: string;
  // In-memory next-seq cache per session, lazily populated by reading the
  // existing folder on first call so we resume correctly after a listener
  // restart.
  private readonly nextSeqs = new Map<string, number>();

  constructor(logsDir?: string) {
    this.logsDir = logsDir ?? process.env.LOGS_DIR ?? DEFAULT_LOGS_DIR;
  }

  private async resolveNextSeq(sessionId: string): Promise<number> {
    const cached = this.nextSeqs.get(sessionId);
    if (cached !== undefined) return cached;
    const dir = sessionDir(this.logsDir, sessionId);
    const files = await fs.readdir(dir).catch(() => [] as string[]);
    let max = 0;
    for (const f of files) {
      const s = seqFromFilename(f);
      if (Number.isFinite(s) && s > max) max = s;
    }
    return max + 1;
  }

  async openVisit(
    sessionId: string,
    targetId: string,
    url: string,
  ): Promise<number> {
    const dir = sessionDir(this.logsDir, sessionId);
    await fs.mkdir(dir, { recursive: true });
    const seq = await this.resolveNextSeq(sessionId);
    this.nextSeqs.set(sessionId, seq + 1);
    const header: VisitHeader = {
      kind: "visit",
      seq,
      targetId,
      url,
      openedAt: new Date().toISOString(),
    };
    await fs.appendFile(
      visitFile(this.logsDir, sessionId, seq, targetId),
      JSON.stringify(header) + "\n",
    );
    return seq;
  }

  async append(
    sessionId: string,
    seq: number,
    targetId: string,
    entry: LogEntry,
  ): Promise<void> {
    await fs.appendFile(
      visitFile(this.logsDir, sessionId, seq, targetId),
      JSON.stringify(entry) + "\n",
    );
  }

  async closeSession(sessionId: string): Promise<void> {
    this.nextSeqs.delete(sessionId);
    await fs
      .rm(sessionDir(this.logsDir, sessionId), { recursive: true, force: true })
      .catch(() => undefined);
  }
}

/**
 * Read events for a session. If `visit` is given, only that visit's events;
 * otherwise events from every visit, ordered by seq. Limit is applied across
 * the merged stream (most recent kept).
 */
export async function readSessionLogs(
  sessionId: string,
  opts: {
    logsDir?: string;
    kind?: "console" | "network";
    limit?: number;
    visit?: number;
  } = {},
): Promise<LogEntry[]> {
  const dir = sessionDir(
    opts.logsDir ?? process.env.LOGS_DIR ?? DEFAULT_LOGS_DIR,
    sessionId,
  );
  const files = await orderedFiles(dir, opts.visit);

  const out: LogEntry[] = [];
  for (const f of files) {
    const stream = createReadStream(path.join(dir, f), { encoding: "utf8" });
    const rl = readline.createInterface({ input: stream, crlfDelay: Infinity });
    try {
      for await (const line of rl) {
        if (!line) continue;
        let entry: LogEntry | VisitHeader;
        try {
          entry = JSON.parse(line) as LogEntry | VisitHeader;
        } catch {
          continue;
        }
        if (entry.kind === "visit") continue; // header line
        if (opts.kind && entry.kind !== opts.kind) continue;
        out.push(entry);
      }
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code !== "ENOENT") throw err;
    }
  }

  if (opts.limit && out.length > opts.limit) {
    return out.slice(-opts.limit);
  }
  return out;
}

/**
 * Read just the visit headers for a session, ordered by seq. Cheap — one
 * line read per file.
 */
export async function readVisits(
  sessionId: string,
  logsDir?: string,
): Promise<VisitHeader[]> {
  const dir = sessionDir(
    logsDir ?? process.env.LOGS_DIR ?? DEFAULT_LOGS_DIR,
    sessionId,
  );
  const files = await orderedFiles(dir);

  const visits: VisitHeader[] = [];
  for (const f of files) {
    const stream = createReadStream(path.join(dir, f), { encoding: "utf8" });
    const rl = readline.createInterface({ input: stream, crlfDelay: Infinity });
    try {
      for await (const line of rl) {
        if (!line) continue;
        try {
          const entry = JSON.parse(line) as { kind?: string };
          if (entry.kind === "visit") visits.push(entry as VisitHeader);
        } catch {
          // ignore
        }
        rl.close();
        break;
      }
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code !== "ENOENT") throw err;
    }
  }

  return visits;
}

async function orderedFiles(dir: string, visit?: number): Promise<string[]> {
  let files: string[];
  try {
    files = await fs.readdir(dir);
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") return [];
    throw err;
  }
  files = files.filter((f) => f.endsWith(".ndjson"));
  if (visit !== undefined) {
    files = files.filter((f) => seqFromFilename(f) === visit);
  }
  // Filenames are zero-padded so lexical sort matches numeric.
  files.sort();
  return files;
}

File: packages/browser-session-mcp/src/logs.rs
//! Per-visit NDJSON event log.
//!
//! Layout: `<logs_dir>/<sessionId>/<seq:05>-<targetId>.ndjson`.
//!
//! Each file starts with a `{"kind":"visit",...}` header line. Subsequent
//! lines are console + network events. The listener opens a new file on every
//! top-level navigation; the MCP server reads them lazily.
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::{
    collections::HashMap,
    path::{Path, PathBuf},
    sync::Arc,
};
use time::{OffsetDateTime, format_description::well_known::Rfc3339};
use tokio::{
    fs,
    io::{AsyncBufReadExt, AsyncWriteExt, BufReader},
    sync::Mutex,
};

const SEQ_PAD: usize = 5;

pub fn default_logs_dir() -> PathBuf {
    std::env::var_os("LOGS_DIR")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("/var/lib/browser-session-mcp/logs"))
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind")]
pub enum LogLine {
    #[serde(rename = "visit")]
    Visit(VisitHeader),
    #[serde(rename = "console")]
    Console(ConsoleEntry),
    #[serde(rename = "network")]
    Network(NetworkEntry),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VisitHeader {
    pub seq: u32,
    #[serde(rename = "targetId")]
    pub target_id: String,
    pub url: String,
    #[serde(rename = "openedAt")]
    pub opened_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConsoleEntry {
    pub t: String,
    #[serde(rename = "type")]
    pub ty: String,
    pub text: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NetworkEntry {
    pub t: String,
    pub method: String,
    pub url: String,
    pub status: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub failure: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LogKind {
    Console,
    Network,
}

#[derive(Clone)]
pub struct LogWriter {
    logs_dir: PathBuf,
    next_seqs: Arc<Mutex<HashMap<String, u32>>>,
}

impl LogWriter {
    pub fn new(logs_dir: PathBuf) -> Self {
        Self {
            logs_dir,
            next_seqs: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    pub async fn open_visit(&self, session_id: &str, target_id: &str, url: &str) -> Result<u32> {
        let dir = session_dir(&self.logs_dir, session_id);
        fs::create_dir_all(&dir)
            .await
            .with_context(|| format!("mkdir -p {}", dir.display()))?;
        let seq = self.next_seq(session_id, &dir).await?;
        let header = VisitHeader {
            seq,
            target_id: target_id.to_string(),
            url: url.to_string(),
            opened_at: now_rfc3339(),
        };
        let line = serde_json::to_string(&LogLine::Visit(header))?;
        let path = visit_file(&self.logs_dir, session_id, seq, target_id);
        append_line(&path, &line).await?;
        Ok(seq)
    }

    pub async fn append_console(
        &self,
        session_id: &str,
        seq: u32,
        target_id: &str,
        entry: ConsoleEntry,
    ) -> Result<()> {
        let path = visit_file(&self.logs_dir, session_id, seq, target_id);
        let line = serde_json::to_string(&LogLine::Console(entry))?;
        append_line(&path, &line).await
    }

    pub async fn append_network(
        &self,
        session_id: &str,
        seq: u32,
        target_id: &str,
        entry: NetworkEntry,
    ) -> Result<()> {
        let path = visit_file(&self.logs_dir, session_id, seq, target_id);
        let line = serde_json::to_string(&LogLine::Network(entry))?;
        append_line(&path, &line).await
    }

    pub async fn close_session(&self, session_id: &str) -> Result<()> {
        self.next_seqs.lock().await.remove(session_id);
        let dir = session_dir(&self.logs_dir, session_id);
        match fs::remove_dir_all(&dir).await {
            Ok(()) => Ok(()),
            Err(err) if err.kind() == std::io::ErrorKind::NotFound => Ok(()),
            Err(err) => Err(err).context(format!("rm -rf {}", dir.display())),
        }
    }

    async fn next_seq(&self, session_id: &str, dir: &Path) -> Result<u32> {
        let mut guard = self.next_seqs.lock().await;
        if let Some(n) = guard.get(session_id).copied() {
            guard.insert(session_id.to_string(), n + 1);
            return Ok(n);
        }
        let mut max = 0u32;
        if let Ok(mut entries) = fs::read_dir(dir).await {
            while let Some(entry) = entries.next_entry().await? {
                if let Some(name) = entry.file_name().to_str() {
                    if let Some(seq) = seq_from_filename(name) {
                        if seq > max {
                            max = seq;
                        }
                    }
                }
            }
        }
        let seq = max + 1;
        guard.insert(session_id.to_string(), seq + 1);
        Ok(seq)
    }
}

pub async fn read_visits(logs_dir: &Path, session_id: &str) -> Result<Vec<VisitHeader>> {
    let dir = session_dir(logs_dir, session_id);
    let files = ordered_files(&dir, None).await?;
    let mut out = Vec::new();
    for f in files {
        let path = dir.join(&f);
        let file = match fs::File::open(&path).await {
            Ok(f) => f,
            Err(err) if err.kind() == std::io::ErrorKind::NotFound => continue,
            Err(err) => return Err(err).context(format!("opening {}", path.display())),
        };
        let mut reader = BufReader::new(file);
        let mut first = String::new();
        if reader.read_line(&mut first).await? == 0 {
            continue;
        }
        let line = first.trim_end_matches(['\r', '\n']);
        if let Ok(LogLine::Visit(v)) = serde_json::from_str::<LogLine>(line) {
            out.push(v);
        }
    }
    Ok(out)
}

pub struct ReadOpts {
    pub kind: Option<LogKind>,
    pub limit: Option<usize>,
    pub visit: Option<u32>,
}

pub enum SessionLogEntry {
    Console(ConsoleEntry),
    Network(NetworkEntry),
}

pub async fn read_session_logs(
    logs_dir: &Path,
    session_id: &str,
    opts: ReadOpts,
) -> Result<Vec<SessionLogEntry>> {
    let dir = session_dir(logs_dir, session_id);
    let files = ordered_files(&dir, opts.visit).await?;
    let mut out: Vec<SessionLogEntry> = Vec::new();
    for f in files {
        let path = dir.join(&f);
        let file = match fs::File::open(&path).await {
            Ok(f) => f,
            Err(err) if err.kind() == std::io::ErrorKind::NotFound => continue,
            Err(err) => return Err(err).context(format!("opening {}", path.display())),
        };
        let mut reader = BufReader::new(file).lines();
        while let Some(line) = reader.next_line().await? {
            if line.is_empty() {
                continue;
            }
            let parsed: LogLine = match serde_json::from_str(&line) {
                Ok(v) => v,
                Err(_) => continue,
            };
            match parsed {
                LogLine::Visit(_) => continue,
                LogLine::Console(c) => {
                    if matches!(opts.kind, Some(LogKind::Network)) {
                        continue;
                    }
                    out.push(SessionLogEntry::Console(c));
                }
                LogLine::Network(n) => {
                    if matches!(opts.kind, Some(LogKind::Console)) {
                        continue;
                    }
                    out.push(SessionLogEntry::Network(n));
                }
            }
        }
    }
    if let Some(limit) = opts.limit {
        if out.len() > limit {
            let start = out.len() - limit;
            out.drain(..start);
        }
    }
    Ok(out)
}

fn sanitize(s: &str) -> String {
    s.chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || c == '_' || c == '-' {
                c
            } else {
                '_'
            }
        })
        .collect()
}

fn session_dir(logs_dir: &Path, session_id: &str) -> PathBuf {
    logs_dir.join(sanitize(session_id))
}

fn visit_file(logs_dir: &Path, session_id: &str, seq: u32, target_id: &str) -> PathBuf {
    session_dir(logs_dir, session_id).join(format!(
        "{:0width$}-{}.ndjson",
        seq,
        sanitize(target_id),
        width = SEQ_PAD
    ))
}

fn seq_from_filename(name: &str) -> Option<u32> {
    let head = name.split('-').next()?;
    head.parse().ok()
}

async fn ordered_files(dir: &Path, visit: Option<u32>) -> Result<Vec<String>> {
    let mut entries = match fs::read_dir(dir).await {
        Ok(e) => e,
        Err(err) if err.kind() == std::io::ErrorKind::NotFound => return Ok(Vec::new()),
        Err(err) => return Err(err).context(format!("read_dir {}", dir.display())),
    };
    let mut out = Vec::new();
    while let Some(entry) = entries.next_entry().await? {
        if let Some(name) = entry.file_name().to_str() {
            if !name.ends_with(".ndjson") {
                continue;
            }
            if let Some(want) = visit {
                if seq_from_filename(name) != Some(want) {
                    continue;
                }
            }
            out.push(name.to_string());
        }
    }
    out.sort();
    Ok(out)
}

async fn append_line(path: &Path, line: &str) -> Result<()> {
    // Single write so O_APPEND keeps the {line + \n} atomic per call —
    // concurrent appenders never tear newlines across records.
    let mut buf = Vec::with_capacity(line.len() + 1);
    buf.extend_from_slice(line.as_bytes());
    buf.push(b'\n');
    let mut f = fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(path)
        .await
        .with_context(|| format!("opening {} for append", path.display()))?;
    f.write_all(&buf)
        .await
        .with_context(|| format!("writing {}", path.display()))?;
    Ok(())
}

fn now_rfc3339() -> String {
    OffsetDateTime::now_utc()
        .format(&Rfc3339)
        .unwrap_or_else(|_| String::new())
}
